### PR TITLE
(PUP-6457) Always attempt remount for out of sync options

### PR DESCRIPTION
**Pull request by HAIL9000**
Prior to this commit, it was possible for puppet to thing that a
mounted device was in sync with it's resource definition when it was
not. This was because if a resource had already been mounted, and a
remount was attempted with bad options, the remount would fail even
though the entry in /etc/fstab would be updated. Puppet was only
checking the resource options against etc/fstab which meant that
after this point, it would believe the mounted object to be in sync
with the puppet resource describing it, even though it was not.

Now, puppet will check the resource's options against both /etc/fstab
and the output of the mount command (on linux operating systems). This
means that it can now detect if the correct object is not mounted and
will attempt a remount.

### CODE REVIEW

File: lib/puppet/provider/mount/parsed.rb [hunk 1]
```diff
@@ -23,7 +23,7 @@
   when "Solaris"
     @fields = [:device, :blockdevice, :name, :fstype, :pass, :atboot, :options]
   else
-    @fields = [:device, :name, :fstype, :options, :dump, :pass]
+    @fields = [:device, :name, :fstype, :options, :dump, :pass, :live_options]
```
Comment (branan) [inline lib/puppet/provider/mount/parsed.rb]: My guess is that the reason our output is broken is because `parsedfile` is actually trying to append the new `live_options` field. I think you want to omit this from the fields variable. We may need to find a way to trick `parsedfile` into accepting the extra data and passing it into the `property_hash` once we remove this, though

Comment (adrienthebo) [inline lib/puppet/provider/mount/parsed.rb]: The implementation of [`Puppet::Provider#set`](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/provider.rb#L543-L547) simply assigns values to `@property_hash`, so we can just call `#set` without having to explicitly add the `live_options` param/value/thing.

Comment (HAIL9000) [inline lib/puppet/provider/mount/parsed.rb]: Thank you!!

File: lib/puppet/type/mount.rb [hunk 2]
```diff
@@ -177,6 +177,29 @@ def syncothers
         appear in fstab. For many platforms this is a comma delimited string.
         Consult the fstab(5) man page for system-specific details."
 
+          def insync?(is)
+            if @resource[:ensure] == :mounted
+              fstab_options = provider.property_hash[:options] || ''
```
Comment (adrienthebo) [inline lib/puppet/type/mount.rb]: `property_hash.fetch(:options, '')` is nice when you want to pick out a value from a hash and provide a default value if the key isn't set.

File: lib/puppet/provider/mount/parsed.rb [hunk 3]
```diff
@@ -264,7 +272,12 @@ def self.mountinstances
     end
     mount_output.each do |line|
       if match = regex.match(line) and name = match.captures.first
-        instances << {:name => name, :mounted => :yes} # Only :name is important here
+        if Facter.value(:kernel) == 'Linux'
```
Comment (adrienthebo) [inline lib/puppet/provider/mount/parsed.rb]: Are we planning on only supporting Linux for remounting on mismatched options, or Solaris and others?

Comment (HAIL9000) [inline lib/puppet/provider/mount/parsed.rb]: Bringing in @branan for feedback on this one. I wasn't quite sure what conclusion we came to, if we should include that work in this ticket or file a new one. It seems like Darwin and SunOS at least would be pretty trivial to add support for but I'm not sure about anything else

File: spec/unit/type/mount_spec.rb [hunk 4]
```diff
@@ -530,6 +530,11 @@ def run_in_catalog(*resources)
       run_in_catalog(resource)
     end
 
+    it "should detect out of sync options" do
+      #provider.property_hash[:live_options] = "foo";
```
Comment (HAIL9000) [inline spec/unit/type/mount_spec.rb]: I'm not sure what to do about this test. It technically works (e.g. it fails without the fix and passes with it) but that is only because if there are no live options retrieved from the output of the mount command, they default to an empty string. I would feel better if we could actually stub that value to something, but whenever I try to add live_options to the property hash of the provider the test freaks out because there's an unrecognized parameter.

**Comment by HAIL9000** [comment]
Hm, some of the failures I saw before are back. They might be related to the switch to using "fetch". Looking into it

File: spec/unit/type/mount_spec.rb (hunk 4 above)
Comment (HAIL9000) [inline spec/unit/type/mount_spec.rb]: This actually has caused me to realize an issue with this solution- if we're running on a non-linux platform no options will be extracted and then in the insync method they will become an empty string meaning that the options will always appear out of sync. Trying to fix this now

**Comment by adrienthebo** [comment]
@HAIL9000 what's the status of this PR? Are the Travis Ci failures known?

**Comment by HAIL9000** [comment]
@adrienthebo Unfortunately the spec test I wrote was only passing because of a bug in my solution and since I fixed it I haven't been able to fix the spec test. What I really need to be able to do is stub `live_options` but I can't figure out a way to do that without Puppet raising an error. So I was just waiting for someone to review this PR and see if they had suggestions (you can see my comments on the spec test above for more details)

**Comment by branan** [comment]
I'm :+1: on this at this point. It doesn't solve all the weird edge cases, but it'll handle the case in the ticket.

We're going to have to put some thought into how to phrase the release notes for this, though

**Comment by branan** [comment]
@HAIL9000 should this be targeted at stable to land in 4.7.1?

**Comment by HAIL9000** [comment]
@branan That seems like a good idea? The customer has been waiting on it for a while. I'm like a little stressed because I FEEL like it SHOULDN'T break anything else but I can't be certain
